elasticdl/python/worker/worker.py

```diff
@@ -814,16 +814,28 @@ def _process_eval_task(self, task):
             tasks.
         """
         logger.info("the evaluation task_id: %d" % task.task_id)
-        eval_info = self._task_data_service.get_validation_dataset(task)
-        if not eval_info:
-            return
-        (eval_dataset, model_version, task_id) = eval_info
-        eval_dataset = self._dataset_fn(
-            eval_dataset,
-            Mode.EVALUATION,
-            self._task_data_service.data_reader.metadata,
-        )
-        eval_dataset = eval_dataset.batch(self._minibatch_size).prefetch(1)
+
+        gen = self._task_data_service.get_validation_dataset_gen(task)
+        if not gen:
+            return None
+
+        @tf.function
+        def create_dataset():
+            eval_dataset = tf.data.Dataset.from_generator(
+                gen, self._task_data_service.data_reader.records_output_types
+            )
+            eval_dataset = self._dataset_fn(
+                eval_dataset,
+                Mode.EVALUATION,
+                self._task_data_service.data_reader.metadata,
+            )
+            eval_dataset = eval_dataset.batch(self._minibatch_size).prefetch(1)
+            return eval_dataset
+
+        with tf.device("/device:cpu:0"):
+            eval_dataset = create_dataset()
```

Comment:
From the logic, the gen is either none or a single instance. I think we can just just create a single eval_dataset instance from gen and assign it as worker's member variable => self.eval_dataset.

Reply:
If we only use a single dataset, we need to determine when the evaluation tasks finish in the dataset pipeline and then continue to execute training tasks.

Reply:
I see, thanks